File: Jinrai.spoon/application_hints_config.lua
local M = {}

local DEFAULT_CONFIG = {
	hotkey = {
		modifiers = nil,
		key = nil,
	},
	windowWaitTimeout = 10,
	apps = {},
	appearance = {
		itemWidth = 220,
		itemHeight = 112,
		gap = 12,
		columns = 3,
		iconSize = 64,
		cornerRadius = 12,
		bgColor = { red = 0.03, green = 0.03, blue = 0.04, alpha = 0.80 },
		dimmedBgColor = { red = 0.03, green = 0.03, blue = 0.04, alpha = 0.30 },
		textColor = { red = 0.96, green = 0.97, blue = 1.00, alpha = 1.00 },
		dimmedTextColor = { red = 0.82, green = 0.84, blue = 0.88, alpha = 0.30 },
		stateColor = { red = 0.40, green = 0.68, blue = 0.98, alpha = 1.00 },
	},
	behavior = {
		callbacks = {
			onError = nil,
		},
	},
	internal = {
		windowHintsKey = nil,
		jinraiModeKey = nil,
		onOpenWindowHints = nil,
		onShowInJinraiMode = nil,
		onStartJinraiMode = nil,
		onSelectInJinraiMode = nil,
		onCancelJinraiMode = nil,
	},
}

local MODIFIER_ORDER = {
	"cmd",
	"alt",
	"ctrl",
	"shift",
	"fn",
}

local MODIFIER_LOOKUP = {
	cmd = true,
	alt = true,
	ctrl = true,
	shift = true,
	fn = true,
}

local MODIFIER_ALIAS_LOOKUP = {
	option = "alt",
}

local DEFAULT_NEW_WINDOW = {
	hotkey = {
		modifiers = { "cmd" },
		key = "n",
	},
	callback = nil,
}

local function isArrayTable(value)
	if type(value) ~= "table" then
		return false
	end
	local maxIndex = 0
	for key, _ in pairs(value) do
		if type(key) ~= "number" or key < 1 or key ~= math.floor(key) then
			return false
		end
		maxIndex = math.max(maxIndex, key)
	end
	for i = 1, maxIndex do
		if value[i] == nil then
			return false
		end
	end
	return true
end

local function deepCopy(value)
	if type(value) ~= "table" then
		return value
	end
	local copied = {}
	for key, item in pairs(value) do
		copied[key] = deepCopy(item)
	end
	return copied
end

local function deepMerge(defaults, overrides)
	if type(defaults) ~= "table" then
		return overrides ~= nil and deepCopy(overrides) or deepCopy(defaults)
	end
	if type(overrides) ~= "table" then
		return overrides ~= nil and deepCopy(overrides) or deepCopy(defaults)
	end
	if isArrayTable(defaults) or isArrayTable(overrides) then
		return deepCopy(overrides)
	end
	local merged = deepCopy(defaults)
	for key, value in pairs(overrides) do
		merged[key] = deepMerge(defaults[key], value)
	end
	return merged
end

local function normalizeKey(value, optionName, allowNil)
	if value == nil and allowNil then
		return nil
	end
	if type(value) ~= "string" or value == "" then
		error(string.format("[jinrai.application_hints] %s must be a non-empty string", optionName))
	end
	return string.upper(value)
end

local function keysConflict(a, b)
	return a == b or string.sub(a, 1, #b) == b or string.sub(b, 1, #a) == a
end

local function normalizeModifiers(modifiers, optionName)
	if type(modifiers) ~= "table" or not isArrayTable(modifiers) then
		error(string.format("[jinrai.application_hints] %s must be an array", optionName))
	end
	if #modifiers == 0 then
		error(string.format("[jinrai.application_hints] %s must not be empty", optionName))
	end
	local lookup = {}
	for index, modifier in ipairs(modifiers) do
		if type(modifier) ~= "string" then
			error(string.format("[jinrai.application_hints] %s[%d] must be a string", optionName, index))
		end
		local normalized = string.lower(modifier)
		normalized = MODIFIER_ALIAS_LOOKUP[normalized] or normalized
		if not MODIFIER_LOOKUP[normalized] then
			error(
				string.format(
					"[jinrai.application_hints] %s[%d] must be one of cmd/alt/ctrl/shift/fn",
					optionName,
					index
				)
			)
		end
		if lookup[normalized] then
			error(string.format("[jinrai.application_hints] %s must not contain duplicate modifiers", optionName))
		end
		lookup[normalized] = true
	end
	local normalized = {}
	for _, modifier in ipairs(MODIFIER_ORDER) do
		if lookup[modifier] then
			table.insert(normalized, modifier)
		end
	end
	return normalized
end

local function normalizeNewWindow(newWindow, optionName)
	if newWindow ~= nil and type(newWindow) ~= "table" then
		error(string.format("[jinrai.application_hints] %s must be a table", optionName))
	end
	local hotkey = deepCopy(DEFAULT_NEW_WINDOW.hotkey)
	local callback = nil
	if newWindow and newWindow.hotkey ~= nil then
		if type(newWindow.hotkey) ~= "table" then
			error(string.format("[jinrai.application_hints] %s.hotkey must be a table", optionName))
		end
		if newWindow.hotkey.modifiers ~= nil then
			hotkey.modifiers = newWindow.hotkey.modifiers
		end
		if newWindow.hotkey.key ~= nil then
			hotkey.key = newWindow.hotkey.key
		end
	end
	if type(hotkey) ~= "table" then
		error(string.format("[jinrai.application_hints] %s.hotkey must be a table", optionName))
	end
	if newWindow then
		callback = newWindow.callback
	end
	if callback ~= nil and type(callback) ~= "function" then
		error(string.format("[jinrai.application_hints] %s.callback must be a function", optionName))
	end
	return {
		hotkey = {
			modifiers = normalizeModifiers(hotkey.modifiers, optionName .. ".hotkey.modifiers"),
			key = string.lower(normalizeKey(hotkey.key, optionName .. ".hotkey.key", false)),
		},
		callback = callback,
	}
end

local function normalizeApps(apps)
	if type(apps) ~= "table" or not isArrayTable(apps) then
		error("[jinrai.application_hints] apps must be an array")
	end
	if #apps == 0 then
		error("[jinrai.application_hints] apps must not be empty")
	end
	local normalized = {}
	for index, app in ipairs(apps) do
		if type(app) ~= "table" then
			error(string.format("[jinrai.application_hints] apps[%d] must be a table", index))
		end
		if type(app.bundleID) ~= "string" or app.bundleID == "" then
			error(string.format("[jinrai.application_hints] apps[%d].bundleID must be a non-empty string", index))
		end
		local key = normalizeKey(app.key, string.format("apps[%d].key", index), false)
		if #key < 1 or #key > 2 then
			error(string.format("[jinrai.application_hints] apps[%d].key must be 1 or 2 characters", index))
		end
		local newWindow = normalizeNewWindow(app.newWindow, string.format("apps[%d].newWindow", index))
		for _, existing in ipairs(normalized) do
			if keysConflict(existing.key, key) then
				error(
					string.format(
						"[jinrai.application_hints] app keys must not duplicate or share a prefix: %s / %s",
						existing.key,
						key
					)
				)
			end
		end
		table.insert(normalized, {
			bundleID = app.bundleID,
			key = key,
			name = app.name,
			newWindow = newWindow,
		})
	end
	return normalized
end

local function positiveNumber(value, optionName)
	if type(value) ~= "number" or value <= 0 or value ~= value then
		error(string.format("[jinrai.application_hints] %s must be > 0", optionName))
	end
	return value
end

function M.build(options)
	options = options or {}
	if type(options) ~= "table" then
		error("[jinrai.application_hints] options must be a table")
	end
	local merged = deepMerge(DEFAULT_CONFIG, options)
	local appearance = merged.appearance
	local apps = normalizeApps(merged.apps)
	local windowHintsKey = normalizeKey(merged.internal.windowHintsKey, "internal.windowHintsKey", true)
	local jinraiModeKey = normalizeKey(merged.internal.jinraiModeKey, "jinrai_mode.triggers.applicationHints.key", true)
	if windowHintsKey and jinraiModeKey and keysConflict(windowHintsKey, jinraiModeKey) then
		error(
			"[jinrai.application_hints] jinrai_mode.triggers.applicationHints.key conflicts with the Window Hints toggle key"
		)
	end
	if windowHintsKey then
		for index, app in ipairs(apps) do
			if keysConflict(app.key, windowHintsKey) then
				error(
					string.format(
						"[jinrai.application_hints] apps[%d].key conflicts with the Window Hints toggle key",
						index
					)
				)
			end
		end
	end
	if jinraiModeKey then
		for index, app in ipairs(apps) do
			if keysConflict(app.key, jinraiModeKey) then
				error(
					string.format(
						"[jinrai.application_hints] apps[%d].key conflicts with jinrai_mode.triggers.applicationHints.key",
						index
					)
				)
			end
		end
	end
	return {
		hotkeyModifiers = merged.hotkey.modifiers,
		hotkeyKey = merged.hotkey.key and string.lower(merged.hotkey.key) or nil,
		windowWaitTimeout = positiveNumber(merged.windowWaitTimeout, "windowWaitTimeout"),
		apps = apps,
		itemWidth = positiveNumber(appearance.itemWidth, "appearance.itemWidth"),
		itemHeight = positiveNumber(appearance.itemHeight, "appearance.itemHeight"),
		gap = positiveNumber(appearance.gap, "appearance.gap"),
		columns = math.floor(positiveNumber(appearance.columns, "appearance.columns")),
		iconSize = positiveNumber(appearance.iconSize, "appearance.iconSize"),
		cornerRadius = positiveNumber(appearance.cornerRadius, "appearance.cornerRadius"),
		bgColor = appearance.bgColor,
		dimmedBgColor = appearance.dimmedBgColor,
		textColor = appearance.textColor,
		dimmedTextColor = appearance.dimmedTextColor,
		stateColor = appearance.stateColor,
		onError = merged.behavior.callbacks.onError,
		windowHintsKey = windowHintsKey,
		jinraiModeKey = jinraiModeKey,
		onOpenWindowHints = merged.internal.onOpenWindowHints,
		onShowInJinraiMode = merged.internal.onShowInJinraiMode,
		onStartJinraiMode = merged.internal.onStartJinraiMode,
		onSelectInJinraiMode = merged.internal.onSelectInJinraiMode,
		onCancelJinraiMode = merged.internal.onCancelJinraiMode,
	}
end

M._test = {
	keysConflict = keysConflict,
	normalizeApps = normalizeApps,
}

return M

File: Sources/JinraiCore/Config/ApplicationHintsConfig.swift
import Foundation

/// Application Hints(アプリランチャー)の設定(元 application_hints_config.lua)
public struct ApplicationHintsConfig: Sendable {
    public struct AppEntry: Sendable {
        /// 起動するアプリの bundle ID(apps[].bundleID)
        public var bundleID: String
        /// 選択キー(1-2文字、大文字化済み)
        public var key: String
        /// 表示名の上書き(nil ならアプリ名を自動解決)
        public var name: String?
        /// 新規ウィンドウ作成のホットキー(起動済みアプリに送出。デフォルト cmd+n)
        public var newWindowModifiers: [String]
        /// 新規ウィンドウ作成のキー(apps[].newWindow.hotkey.key)
        public var newWindowKey: String
        /// 宣言的な新規ウィンドウ URL(元 callback の代替。指定時は最優先で open)
        public var newWindowURL: String?
    }

    /// Application Hints を直接開くホットキーの修飾キー(hotkey.modifiers)
    public var hotkeyModifiers: [String]
    /// Application Hints を直接開くキー(hotkey.key。nil で無効)
    public var hotkeyKey: String?
    /// 新規ウィンドウが現れるまで待つ最大時間(秒。windowWaitTimeout)
    public var windowWaitTimeout: Double
    /// 対象アプリの一覧(apps。1件以上必須)
    public var apps: [AppEntry]

    /// 1行に表示するアプリ数(appearance.columns)
    public var columns: Int
    /// アプリ項目の幅(appearance.itemWidth)
    public var itemWidth: Double
    /// アプリ項目の高さ(appearance.itemHeight)
    public var itemHeight: Double
    /// アプリ項目同士の間隔(appearance.gap)
    public var gap: Double
    /// アプリアイコンの大きさ(appearance.iconSize)
    public var iconSize: Double
    /// アプリ項目背景の角丸(appearance.cornerRadius)
    public var cornerRadius: Double
    /// 通常時の背景色(appearance.bgColor)
    public var bgColor: ConfigColor
    /// 候補から外れた項目の背景色(appearance.dimmedBgColor)
    public var dimmedBgColor: ConfigColor
    /// 通常時の文字色(appearance.textColor)
    public var textColor: ConfigColor
    /// 候補から外れた項目の文字色(appearance.dimmedTextColor)
    public var dimmedTextColor: ConfigColor
    /// OPEN または NEW の状態表示色(appearance.stateColor)
    public var stateColor: ConfigColor
    /// 入力済みプレフィックス文字の色(Window Hints の hint.key.keyHighlightColor 相当)
    public var keyHighlightColor: ConfigColor

    /// Window Hints へ戻る遷移キー(init 相当の結線で注入)
    public var windowHintsKey: String?
    /// 表示中に JinraiMode を開始するキー(jinraiMode.triggers.applicationHints.key)
    public var jinraiModeKey: String?
}

public enum ApplicationHintsConfigBuilder {
    static var defaults: [String: Any] {
        [
            "windowWaitTimeout": 10,
            "appearance": [
                "itemWidth": 220,
                "itemHeight": 112,
                "gap": 12,
                "columns": 3,
                "iconSize": 64,
                "cornerRadius": 12,
                "bgColor": ["red": 0.03, "green": 0.03, "blue": 0.04, "alpha": 0.80],
                "dimmedBgColor": ["red": 0.03, "green": 0.03, "blue": 0.04, "alpha": 0.30],
                "textColor": ["red": 0.96, "green": 0.97, "blue": 1.00, "alpha": 1.00],
                "dimmedTextColor": ["red": 0.82, "green": 0.84, "blue": 0.88, "alpha": 0.30],
                "stateColor": ["red": 0.40, "green": 0.68, "blue": 0.98, "alpha": 1.00],
                "keyHighlightColor": ["red": 0.84, "green": 0.84, "blue": 0.86, "alpha": 0.35],
            ],
        ]
    }

    /// 完全一致 or 一方が他方の接頭辞なら衝突(元 keysConflict)
    static func keysConflict(_ a: String, _ b: String) -> Bool {
        ConfigKeyDescriptor.sequence(a).conflicts(with: ConfigKeyDescriptor.sequence(b))
    }

    public static func build(
        _ options: [String: Any], windowHintsKey: String? = nil
    ) throws -> ApplicationHintsConfig {
        let merged = ConfigDict(
            DeepMerge.merge(defaults: defaults, overrides: options), context: "applicationHints")

        guard let rawApps = merged.value("apps") as? [[String: Any]], !rawApps.isEmpty else {
            throw ConfigError("[jinrai.applicationHints] apps は1件以上必要です")
        }

        var apps: [ApplicationHintsConfig.AppEntry] = []
        for (index, rawApp) in rawApps.enumerated() {
            guard let bundleID = rawApp["bundleID"] as? String, !bundleID.isEmpty else {
                throw ConfigError(
                    "[jinrai.applicationHints] apps[\(index)].bundleID は必須です")
            }
            guard let rawKey = rawApp["key"] as? String, !rawKey.isEmpty else {
                throw ConfigError("[jinrai.applicationHints] apps[\(index)].key は必須です")
            }
            let key = rawKey.uppercased()
            guard (1...2).contains(key.count) else {
                throw ConfigError(
                    "[jinrai.applicationHints] apps[\(index)].key は1〜2文字です: \(rawKey)")
            }

            var modifiers = ["cmd"]
            var newWindowKey = "n"
            var url: String?
            if let newWindow = rawApp["newWindow"] as? [String: Any] {
                url = newWindow["url"] as? String
                if let hotkey = newWindow["hotkey"] as? [String: Any] {
                    if let mods = hotkey["modifiers"] as? [String], !mods.isEmpty {
                        modifiers = mods
                    }
                    if let k = hotkey["key"] as? String, !k.isEmpty {
                        newWindowKey = k.lowercased()
                    }
                }
            }

            apps.append(
                .init(
                    bundleID: bundleID,
                    key: key,
                    name: rawApp["name"] as? String,
                    newWindowModifiers: modifiers,
                    newWindowKey: newWindowKey,
                    newWindowURL: url
                ))
        }

        // キー衝突検証: apps 間
        for i in apps.indices {
            for j in apps.indices where i < j {
                if keysConflict(apps[i].key, apps[j].key) {
                    throw ConfigError(
                        "[jinrai.applicationHints] キー '\(apps[i].key)' と '\(apps[j].key)' が衝突します"
                    )
                }
            }
        }
        // windowHintsKey との衝突
        if let windowHintsKey = windowHintsKey?.uppercased() {
            let windowHintsDescriptor = ConfigKeyDescriptor.keyName(windowHintsKey)
            for app in apps
            where ConfigKeyDescriptor.sequence(app.key).conflicts(with: windowHintsDescriptor) {
                throw ConfigError(
                    "[jinrai.applicationHints] キー '\(app.key)' が Window Hints 遷移キー '\(windowHintsKey)' と衝突します"
                )
            }
        }

        func positiveDouble(_ path: String, _ fallback: Double) throws -> Double {
            let value = merged.double(path) ?? fallback
            guard value > 0 else {
                throw ConfigError("[jinrai.applicationHints] \(path) は正の数が必要です")
            }
            return value
        }

        return ApplicationHintsConfig(
            hotkeyModifiers: merged.stringArray("hotkey.modifiers") ?? [],
            hotkeyKey: merged.string("hotkey.key"),
            windowWaitTimeout: try positiveDouble("windowWaitTimeout", 10),
            apps: apps,
            columns: max(1, merged.int("appearance.columns") ?? 3),
            itemWidth: try positiveDouble("appearance.itemWidth", 220),
            itemHeight: try positiveDouble("appearance.itemHeight", 112),
            gap: merged.double("appearance.gap") ?? 12,
            iconSize: try positiveDouble("appearance.iconSize", 64),
            cornerRadius: merged.double("appearance.cornerRadius") ?? 12,
            bgColor: merged.color("appearance.bgColor")
                ?? ConfigColor(red: 0.03, green: 0.03, blue: 0.04, alpha: 0.80),
            dimmedBgColor: merged.color("appearance.dimmedBgColor")
                ?? ConfigColor(red: 0.03, green: 0.03, blue: 0.04, alpha: 0.30),
            textColor: merged.color("appearance.textColor")
                ?? ConfigColor(red: 0.96, green: 0.97, blue: 1.00, alpha: 1.00),
            dimmedTextColor: merged.color("appearance.dimmedTextColor")
                ?? ConfigColor(red: 0.82, green: 0.84, blue: 0.88, alpha: 0.30),
            stateColor: merged.color("appearance.stateColor")
                ?? ConfigColor(red: 0.40, green: 0.68, blue: 0.98, alpha: 1.00),
            keyHighlightColor: merged.color("appearance.keyHighlightColor")
                ?? ConfigColor(red: 0.84, green: 0.84, blue: 0.86, alpha: 0.35),
            windowHintsKey: windowHintsKey?.uppercased(),
            jinraiModeKey: nil  // RootConfigBuilder が jinrai_mode.triggers から注入
        )
    }
}
